Test pipeline parallelism works with DDP.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49577 Add benchmakr for torch.distributed.pipeline.sync.Pipe
* #48638 Improve documentation for pipeline parallelism.
* #48630 Cleanup APIs for pipeline parallelism.
* **#48470 Test pipeline parallelism works with DDP.**
* #49419 Use store based barrier in init_process_group.

Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

pritamdamania committed:
Test pipeline parallelism works with DDP.

Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/pipe_with_ddp_test.py:
```diff
@@ -0,0 +1,101 @@
+    def world_size(self) -> int:
+        return 4
+
+    @skip_if_lt_x_gpu(4)
```
IIRC, we will need to run ci-all to trigger multigpu tests to cover this

pritamdamania87 replied:
Tests are passing for multigpu: https://app.circleci.com/pipelines/github/pytorch/pytorch/250277/workflows/77bc3b3b-2a7c-46f7-b0ea-e06a496d1e4b/jobs/9579475, PR: https://github.com/pytorch/pytorch/pull/49302

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/pipe_with_ddp_test.py:
```diff
@@ -0,0 +1,101 @@
+            # Check grads
+            output = [torch.empty_like(fc1.weight.grad), torch.empty_like(fc1.weight.grad)]
+            pg.allgather(output, fc1.weight.grad)
+            torch.cuda.synchronize(self.rank)
```
curious, do we need to synchronize here? Will the compare op below be automatically inserted to the current stream?

pritamdamania87 replied:
Didn't need this and removed it.

mrshenli commented on torch/testing/_internal/distributed/pipe_with_ddp_test.py:
```diff
@@ -0,0 +1,101 @@
+from torch.nn.parallel import DistributedDataParallel
+import torch.distributed as dist
+from torch import nn
+import torch
+from torch.testing._internal.dist_utils import INIT_METHOD_TEMPLATE, dist_init
+from torch.testing._internal.common_distributed import (
+    requires_gloo,
+    requires_nccl,
+    skip_if_lt_x_gpu,
+    skip_if_rocm,
+)
+from torch.testing._internal.distributed.rpc.rpc_agent_test_fixture import (
+    RpcAgentTestFixture,
+)
+from torch.distributed._pipeline.sync import Pipe
+import unittest
```
nit: shall we sort the import block a bit?

mrshenli reviewed (approved):
LGTM! Thanks for adding this!

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Test pipeline parallelism works with DDP."


Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

[ghstack-poisoned]
